internal/server/filesystem/server.go

```diff
@@ -2,24 +2,227 @@ package filesystem
+	}
+
+	if isUNCPathWindows(path) {
+		return fmt.Errorf("Unsupported UNC path prefix: %s", path)
+	}
+
+	if containsInvalidCharactersWindows(path) {
+		return fmt.Errorf("Path contains invalid characters: %s", path)
+	}
+
+	if isAbsWindows(path) && !strings.HasPrefix(path, prefix) {
+		return fmt.Errorf("Absolute path: %s is not within context path: %s", path, prefix)
+	}
+
+	return nil
+}
+
+func (s *Server) abs(pathCtx internal.PathContext, path string) (string, error) {
+	if isAbsWindows(path) {
+		return path, nil
+	}
+	prefix := ""
+	if pathCtx == internal.PLUGIN {
+		prefix = s.kubeletCSIPluginsPath
+	} else if pathCtx == internal.CONTAINER {
+		prefix = s.kubeletPodPath
+	} else {
+		return "", fmt.Errorf("Invalid PathContext: %v", pathCtx)
+	}
+	return prefix + "\\" + path, nil
+}
 
 // PathExists checks if the given path exists on the host.
 func (s *Server) PathExists(ctx context.Context, request *internal.PathExistsRequest, version apiversion.Version) (*internal.PathExistsResponse, error) {
-	// FIXME: actually implement this!
+	err := s.validatePathWindows(request.Context, request.Path)
+	if err != nil {
+		return &internal.PathExistsResponse{
+			Error: err.Error(),
+		}, nil
+	}
+
+	path, err := s.abs(request.Context, request.Path)
+	if err != nil {
+		return &internal.PathExistsResponse{
+			Error: err.Error(),
+		}, nil
+	}
+
+	exists, err := s.hostAPI.PathExists(path)
 	return &internal.PathExistsResponse{
-		Success: false,
-		CmdletError: &api.CmdletError{
-			CmdletName: "dummy",
-			Code:       12,
-			Message:    "hey there " + request.Path,
-		},
+		Error:  err.Error(),
+		Exists: exists,
+	}, nil
+}
+
+func (s *Server) Mkdir(ctx context.Context, request *internal.MkdirRequest, version apiversion.Version) (*internal.MkdirResponse, error) {
+	err := s.validatePathWindows(request.Context, request.Path)
+	if err != nil {
+		return &internal.MkdirResponse{
+			Error: err.Error(),
+		}, nil
+	}
+
+	path, err := s.abs(request.Context, request.Path)
+	if err != nil {
+		return &internal.MkdirResponse{
+			Error: err.Error(),
+		}, nil
+	}
+
+	err = s.hostAPI.Mkdir(path)
+	if err != nil {
+		return &internal.MkdirResponse{
+			Error: err.Error(),
+		}, nil
+	}
+
+	return &internal.MkdirResponse{
+		Error: "",
+	}, nil
+}
+
+func (s *Server) Rmdir(ctx context.Context, request *internal.RmdirRequest, version apiversion.Version) (*internal.RmdirResponse, error) {
+	err := s.validatePathWindows(request.Context, request.Path)
+	if err != nil {
+		return &internal.RmdirResponse{
+			Error: err.Error(),
+		}, nil
+	}
+	err = s.hostAPI.Rmdir(request.Path)
+	if err != nil {
+		return &internal.RmdirResponse{
+			Error: err.Error(),
+		}, nil
+	}
+	return &internal.RmdirResponse{
+		Error: "",
+	}, nil
+}
+
+func (s *Server) LinkPath(ctx context.Context, request *internal.LinkPathRequest, version apiversion.Version) (*internal.LinkPathResponse, error) {
+	err := s.validatePathWindows(internal.CONTAINER, request.SourcePath)
```

Comment:
why here to use internal.CONTAINER directly instead of request.Context?

Reply:
In general we want to constrain the operations within FS contexts as much as possible (as commented in [KEP](https://github.com/kubernetes/enhancements/pull/1184#discussion_r308935813)). For Mkdir/Rmdir, we need to allow both as explained [above](https://github.com/kubernetes-csi/csi-proxy/pull/11#discussion_r339923451). LinkPath however is only needed for linking the node-global staging paths under PLUGIN context to pod-local staging paths under CONTAINER context.